### ISSUE TITLE
Conversion to tensor required when predicting using TFP distribution layers

### ISSUE THREAD

**Issue by skeydan** [rstudio/keras#793]
Working with TFP distribution layers, predictions have to be made using

```
model(testdata)
```

In `fit` we can work with R types (matrices/arrays), but for prediction we need to convert to tensors.

E.g.

```
library(tensorflow)
tf$enable_v2_behavior()
library(tfprobability)
library(keras)

x <- 1.1:10 %>% as.matrix()
y <- 1.1:10 %>% as.matrix()
x_test <-x

model <- keras_model_sequential() %>%
  layer_dense(units = 2) %>%
  layer_distribution_lambda(
    function(x) tfd_normal(loc = x[ , 1, drop = FALSE],
                           scale = 1e-3 + tf$math$softplus(0.05 * x[ , 2, drop = FALSE]))
    )

negloglik <- function(y, model) -(model %>% tfd_log_prob(y))
model %>% compile(optimizer = optimizer_adam(lr = 0.01), loss = negloglik)
model %>% fit(x, y, epochs = 1)

yhat <- model(tf$constant(x_test))
# yhat <- model(x_test) error
```

If we do 

`yhat <- model(x_test) ` 

instead we get

```
> yhat <- model(x_test)
Error: Invalid input to layer function (must be a model or a tensor)
> traceback()
5: stop("Invalid input to layer function (must be a model or a tensor)", 
       call. = FALSE) at layers-core.R#549
4: stop_with_invalid_layer() at layers-core.R#526
3: compose_layer.default(object, x) at layers-core.R#522
2: compose_layer(object, x) at model.R#191
1: model(x_test)
```

We might consider changing the validation here or do the conversion for the user (putting this here for later follow-up)

**Comment by dfalbel** [comment]
What do you think about adding a `predict` method for tfprobability models?

**Comment by skeydan** [comment]
excellent idea!!

However, they have the same class hierarchy as other `keras` models, so probably `keras:::predict.keras.engine.training.Model` should handle this...

Actually now I'm thinking - perhaps TFP will implement `model.predict()` for these cases too?
I think we could see about the outcome of 
https://github.com/tensorflow/probability/issues/427 
(which I created just now)

**Comment by dfalbel** [comment]
Maybe we could add something like `predict_distribution` like we have `predict_proba` and `predict_class` ?

**Comment by skeydan** [comment]
yeah, but these exist in Python, too, so these functions would be a  bit inhomogeneous...

On the other hand, fixing the above `model(data)` to work with R objects would be a bit hacky I guess.
That it _does_ work with tensors is because we're landing in 

```
compose_layer.python.builtin.object <- function(object, layer) {
  if (is.function(layer))
    layer(object)
  else
    stop_with_invalid_layer()
}
```

and then, the model gets called because it _is_ a function... But somehow this call has nothing to do with layer composition...

BTW what I still don't get is why we're landing in `compose_layer` at all?

But in that respect, yes perhaps we should add a function of our own, that under the hood does `model(tf$convert_to_tensor(data))`?
Not sure about the name though (in the sense that I don't know if this should always be called when using TFP-keras models, or just when those end in a `DistributionLambda`). This is why I was hoping they'd perhaps make `predict` work on their side, then we could just use that.

```
compose_layer.python.builtin.object <- function(object, layer) {
  if (is.function(layer))
    layer(object)
  else
    stop_with_invalid_layer()
}
```

and then, the model gets called because it _is_ a function... But somehow this call has nothing to do with layer composition...

BTW what I still don't get is why we're landing in `compose_layer` at all?

But in that respect, yes perhaps we should add a function of our own, that under the hood does `model(tf$convert_to_tensor(data))`?
Not sure about the name though (in the sense that I don't know if this should always be called when using TFP-keras models, or just when those end in a `DistributionLambda`). This is why I was hoping they'd perhaps make `predict` work on their side, then we could just use that.

**Comment by dfalbel** [comment]
Fixed with #806
